applications/main/nfc/plugins/supported_cards/charliecard.c

```diff
@@ -0,0 +1,1032 @@
+    if(furi_hal_rtc_is_flag_set(FuriHalRtcFlagDebug)) {
+        furi_string_cat_printf(out, "%s%u%s%u", sep, trip.g_flag, sep, trip.f_flag);
+    }
+}
+
+static bool charliecard_parse(const NfcDevice* device, FuriString* parsed_data) {
+    furi_assert(device);
+
+    const MfClassicData* data = nfc_device_get_data(device, NfcProtocolMfClassic);
+
+    bool parsed = false;
+
+    do {
+        // Verify card type
+        if(data->type != MfClassicType1k) break;
+
+        // Verify key
+        const uint8_t verify_sector = 3; // arbitrary
+        const MfClassicSectorTrailer* sec_tr =
+            mf_classic_get_sector_trailer_by_sector(data, verify_sector);
+
+        const uint64_t key =
+            bit_lib_bytes_to_num_be(sec_tr->key_a.data, COUNT_OF(sec_tr->key_a.data));
+        if(key != charliecard_1k_keys[verify_sector].a) break;
+
+        // TODO: Verify add'l?
+
+        const enum CharlieActiveSector active_sec_enum = get_active_sector(data);
+        const uint8_t active_sector = (active_sec_enum == CHARLIE_ACTIVE_SECTOR_2) ? 2 : 3;
+
+        furi_string_cat_printf(parsed_data, "\e#CharlieCard");
+
+        size_t uid_len = 0;
+        const uint8_t* uid = mf_classic_get_uid(data, &uid_len);
+        uint32_t card_number = bit_lib_bytes_to_num_be(uid, 4);
+        furi_string_cat_printf(parsed_data, "\nSerial: 5-%lu", card_number);
+
+        // Money fare = money_parse(data, active_sector, 0, 4);
+        Money bal = money_parse(data, active_sector, 1, 5);
+        furi_string_cat_printf(parsed_data, "\nBal: ");
+        money_format_cat(parsed_data, bal);
+
+        const uint16_t type = type_parse(data);
+        furi_string_cat_printf(parsed_data, "\nType: ");
+        type_format_cat(parsed_data, type);
+
+        const uint16_t n_trips = n_uses(data, active_sec_enum);
+        furi_string_cat_printf(parsed_data, "\nTrip Count: %u", n_trips);
+
+        const DateTime iss = date_parse(data, active_sector, 0, 6);
+        furi_string_cat_printf(parsed_data, "\nIssued: ");
+        locale_format_dt_cat(parsed_data, &iss);
+
+        const DateTime e_v = end_validity_parse(data, active_sec_enum);
+        furi_string_cat_printf(parsed_data, "\nExpiry: ");
+        locale_format_dt_cat(parsed_data, &e_v);
+
+        /*
+        const DateTime exp = expiry(iss);
+        furi_string_cat_printf(parsed_data, "\nExp: ");
+        locale_format_dt_cat(parsed_data, &exp);*/
+
+        DateTime last = date_parse(data, active_sector, 0, 1);
+        furi_string_cat_printf(parsed_data, "\nExpired: %s", expired(e_v, last) ? "Yes" : "No");
+
+        const Trip* trips = trips_parse(data);
+        furi_string_cat_printf(parsed_data, "\nTransactions:");
+        for(size_t i = 0; i < CHARLIE_N_TRIP_HISTORY; i++) {
+            furi_string_cat_printf(parsed_data, "\n");
+            trip_format_cat(parsed_data, trips[i]);
+            furi_string_cat_printf(parsed_data, "\n");
+        }
+        // free(trips);
```

Comment:
this is the only issue i can see, memory-management-wise. `trips_parse()` allocates the buffer, it is used here, and then it's not free'd. simply uncommenting that line should work, is there a particular reason why it was commented?

Reply:
Ope, good catch! I had commented that out as it wasn't compiling with it uncommented, and simply forgot to go back and check why I was having compile-time issues to begin with. Turns out I simply had an unnecessary `const` on line 999: `const Trip* trips = trips_parse(data);` [sic]. Removing the `const` and uncommenting the `free()` makes the compile-time checks happy, and should fix the above.